daemon: restore: clear net state for stopped containers

akerouanton opened the pull request:
**- What I did**

**daemon: restore: clear net state for stopped containers**

When the daemon crashes, the host unexpectedly reboot, or the daemon restarts with live-restore enabled, running containers might stop and the on-disk state for containers might diverge from reality. All these situations are currently handled by the daemon's `restore` method.

That method calls `daemon.Cleanup()` for all the dead containers. In turn, `Cleanup` calls `daemon.releaseNetwork()`. However, this last method won't do anything because it expects the `netController` to be initialized when it's called. That's not the case in the `restore` code path -- the `netController` is initialized _after_ cleaning up dead containers.

There's a chicken-egg problem here, and fixing that would require some important architectural changes (eg. change the way libnet's controller is initialized).

Since `releaseNetwork()` early exits, dead containers won't ever have their networking state cleaned. This led to bugs in Docker Desktop among other things.

Fix that by calling `releaseNetwork` after initializing the `netController`.

**daemon: releaseNetwork: clear SandboxID, SandboxKey**

When the container stops or during `restore`, `daemon.releaseNetwork` is used to clear all net-related state carried by a container. However, the fields `SandboxID` and `SandboxKey` are never cleared. On the next start, these fields will be replaced with new values. There's no point in preserving these data since they became invalid as soon as the container stopped.

**- How to verify it**

```console
# Start the daemon with --live-restore
$ docker run -d --name c0 -p 80:80 traefik/whoami
$ docker inspect c0
# Kill the daemon
# Kill the container process
# Restart the daemon
$ docker inspect c0
```

**- Description for the changelog**

```markdown changelog
- Clear the networking state of all stopped / dead containers during daemon start-up.
```

akerouanton committed:
daemon: restore: clear net state for stopped containers

When the daemon crashes, the host unexpectedly reboot, or the daemon
restarts with live-restore enabled, running containers might stop and the
on-disk state for containers might diverge from reality. All these
situations are currently handled by the daemon's `restore` method.

That method calls `daemon.Cleanup()` for all the dead containers. In
turn, `Cleanup` calls `daemon.releaseNetwork()`. However, this last
method won't do anything because it expects the `netController` to be
initialized when it's called. That's not the case in the `restore` code
path -- the `netController` is initialized _after_ cleaning up dead
containers.

There's a chicken-egg problem here, and fixing that would require some
important architectural changes (eg. change the way libnet's controller
is initialized).

Since `releaseNetwork()` early exits, dead containers won't ever have
their networking state cleaned. This led to bugs in Docker Desktop,
among other things.

Fix that by calling `releaseNetwork` after initializing the
`netController`.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

akerouanton committed:
daemon: releaseNetwork: clear SandboxID, SandboxKey

When the container stops or during `restore`, `daemon.releaseNetwork` is
used to clear all net-related state carried by a container. However, the
fields `SandboxID` and `SandboxKey` are never cleared. On the next start,
these fields will be replaced with new values. There's no point in
preserving these data since they became invalid as soon as the container
stopped.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

robmry commented on daemon/daemon.go:
```diff
@@ -591,6 +591,11 @@ func (daemon *Daemon) restore(cfg *configStore) error {
 	for _, c := range containers {
```
The comment above now only applies to the `if` branch ... could move it into there?

robmry commented on daemon/daemon.go:
```diff
@@ -591,6 +591,11 @@ func (daemon *Daemon) restore(cfg *configStore) error {
 		if sb, err := daemon.netController.SandboxByID(c.NetworkSettings.SandboxID); err == nil {
 			c.NetworkSettings.Ports = getPortMapInfo(sb)
+		} else {
+			// If the sandbox for this container don't exist anymore, that's
```
don't -> doesn't

robmry reviewed (approved):
Just comments on comments!

thaJeztah commented:
Can we add a test for this somewhere, or is that complicated?


Also; ISTR there was a regression point for this (i recall some patch had to be added in docker desktop - possible after a regression?) if so; do you know what versions had this issue, and if this is back-portable to those versions?

thaJeztah commented on daemon/container_operations.go:
```diff
@@ -1008,6 +1007,9 @@ func (daemon *Daemon) releaseNetwork(container *container.Container) {
 		return
 	}
 
+	container.NetworkSettings.SandboxID = ""
+	container.NetworkSettings.SandboxKey = ""
```
I see we don't reset `container.NetworkSettings.SandboxKey` if we find that `container.NetworkSettings.SandboxID` is empty. Could there be any situations where it's non-empty in that case (and would _not_ resetting it cause issues in that situation?)

thaJeztah replied:
(considering if we should move the resetting above the `if sid == ""` branch above if there could be such situations, and if those would be problematic.

akerouanton replied:
The only other place where `SandboxID` is written is here:

https://github.com/moby/moby/blob/076c976e28cac7802835e944ae03cee2662cbea2/daemon/container_operations.go#L1160-L1164

Before this commit, `SandboxID` was never cleared. So, the purpose of this condition is to prevent calling `releaseNetwork` twice or more on the same container. I think it's safe to not check them / clear them separately.

robmry commented on integration/container/daemon_test.go:
```diff
@@ -50,3 +51,50 @@ func TestContainerKillOnDaemonStart(t *testing.T) {
+	assert.Assert(t, inspect.NetworkSettings.SandboxKey != "")
+	assert.Assert(t, inspect.NetworkSettings.Ports["80/tcp"] != nil)
+
+	assert.NilError(t, d.Kill())
+	d.Start(t)
```
I'm not sure if this is really an issue - but `Kill` only sends the signal, it doesn't hang around for the process to disappear. So, I wonder if there's a possible race between the OS cleaning things up and the `d.Start()`?

akerouanton committed:
testutil/daemon: Wait() until the daemon is Kill()'ed

`Daemon.Kill()` was sending a SIGKILL to the daemon process but wasn't
waiting until the process was really killed. While the race window is
really small, better safe than sorry.

Signed-off-by: Albin Kerouanton <albinker@gmail.com>

thaJeztah commented on testutil/daemon/daemon.go:
```diff
@@ -621,6 +621,11 @@ func (d *Daemon) Kill() error {
 		return err
 	}
 
+	_, err := d.cmd.Process.Wait()
```
Will this affect other tests that don't require it to be gone  (ie should this be optional?)

akerouanton replied:
SIGKILL is sent to the daemon process by this method. So I don't see how other tests couldn't require it not to be gone.

thaJeztah replied:
Yeah, maybe it's fine; the case I had in mind would be a test spinning up a daemon, killing one, and creating a new one (without caring about the previous one - not sure if the wait would cause our CI to take longer)

akerouanton replied:
If another daemon is started with the exact same config (eg. same data dirs), there's a short window of time where both daemons might collide. So overall, even if it adds a few ms of delay, I think it's a net gain. I even removed a `time.Sleep` that was presumably added to avoid that sort of bug.

corhere commented on daemon/daemon.go:
```diff
@@ -583,14 +583,20 @@ func (daemon *Daemon) restore(cfg *configStore) error {
-	// will be running. Replace the restored mappings in NetworkSettings with the
-	// current state so that the problem is visible in 'inspect'.
 	for _, c := range containers {
 		if sb, err := daemon.netController.SandboxByID(c.NetworkSettings.SandboxID); err == nil {
+			// If port-mapping failed during live-restore of a container, perhaps because
+			// a host port that was previously mapped to a container is now in-use by some
+			// other process - ports will not be mapped for the restored container, but it
+			// will be running. Replace the restored mappings in NetworkSettings with the
+			// current state so that the problem is visible in 'inspect'.
 			c.NetworkSettings.Ports = getPortMapInfo(sb)
+		} else {
+			// If this container's sandbox doesn't exist anymore, that's because
+			// the netController gc'ed it during its initialization. In that case,
+			// we need to clear all the network-related state carried by that
+			// container.
+			daemon.releaseNetwork(c)
 		}
```
https://go.dev/wiki/CodeReviewComments#indent-error-flow
```suggestion
		sb, err := daemon.netController.SandboxByID(c.NetworkSettings.SandboxID)
		if err != nil {
			// If this container's sandbox doesn't exist anymore, that's because
			// the netController gc'ed it during its initialization. In that case,
			// we need to clear all the network-related state carried by that
			// container.
			daemon.releaseNetwork(c)
			continue
		}
		// If port-mapping failed during live-restore of a container, perhaps because
		// a host port that was previously mapped to a container is now in-use by some
		// other process - ports will not be mapped for the restored container, but it
		// will be running. Replace the restored mappings in NetworkSettings with the
		// current state so that the problem is visible in 'inspect'.
		c.NetworkSettings.Ports = getPortMapInfo(sb)
```

corhere commented on testutil/daemon/daemon.go:
```diff
@@ -621,6 +621,11 @@ func (d *Daemon) Kill() error {
 	}
 
+	_, err := d.cmd.Process.Wait()
+	if err != nil && !strings.Contains(err.Error(), "no child processes") {
```
```suggestion
	if err != nil && !errors.Is(err, syscall.ECHILD) {
```
FYI, [`ECHILD` is defined on Windows.](https://pkg.go.dev/syscall?GOOS=windows#:~:text=EBFONT%0A%09EBUSY%0A%09ECANCELED-,ECHILD,-ECHRNG%0A%09ECOMM%0A%09ECONNABORTED)

corhere reviewed (approved):
LGTM, style nits aside.

thaJeztah reviewed (approved):
LGTM

thaJeztah commented:
@akerouanton do you want to address Cory's comments from above before this is merged?

akerouanton commented:
@thaJeztah I just force-pushed. We can merge once CI is green.

thaJeztah commented:
@akerouanton some failure;

```
56.11 daemon/daemon.go:593:26: not enough arguments in call to daemon.releaseNetwork
56.11 	have (*"github.com/docker/docker/container".Container)
56.11 	want ("context".Context, *"github.com/docker/docker/container".Container)
```

corhere reviewed (approved):
This PR gets my context-plumbing stamp of approval ✅

vvoland commented:
The ppc64le job is also stuck on other PRs, so it's unrelated, and the rest CI is green.
I'll get this in.
